tsl/src/gapfill/exec.c

```diff
@@ -244,6 +293,198 @@ is_simple_expr(Expr *node)
+
+	/*
+	 * if the second argument to time_bucket_gapfill is not a column reference
+	 * we cannot match WHERE clause to the time column, we don't check for Var
+	 * directly though cause it might be wrapped in a cast
+	 */
+	vars = pull_var_clause(lsecond(func->args), 0);
+	if (list_length(vars) != 1)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid time_bucket_gapfill argument: ts needs to refer to column if no "
+						"start or finish is supplied"),
+				 errhint(
+					 "You can either pass start and finish as arguments or in the WHERE clause")));
+
+	foreach (lc, (List *) jt->quals)
+	{
+		OpExpr *opexpr;
+		Var *var;
+		Expr *expr;
+		Oid op;
+		int64 value;
+
+		if (!IsA(lfirst(lc), OpExpr))
+			continue;
+
+		opexpr = lfirst(lc);
+
+		if (IsA(linitial(opexpr->args), Var))
+		{
+			var = linitial(opexpr->args);
+			expr = lsecond(opexpr->args);
+			op = opexpr->opno;
+		}
+		else if (IsA(lsecond(opexpr->args), Var))
+		{
+			var = lsecond(opexpr->args);
+			expr = linitial(opexpr->args);
+			op = get_commutator(opexpr->opno);
+		}
+		else
+			continue;
+
+		if (!op_in_opfamily(op, tce->btree_opf))
+			continue;
+
+		/*
+		 * only allow simple expressions because Params have not been set up
+		 * at this stage and Vars will not work either because we execute in
+		 * separate execution context
+		 */
+		if (!is_simple_expr(expr) || !var_equal(linitial(vars), var))
+			continue;
+
+		get_op_opfamily_properties(op, tce->btree_opf, false, &strategy, &lefttype, &righttype);
+
+		if (boundary == GAPFILL_START && strategy != BTGreaterStrategyNumber &&
+			strategy != BTGreaterEqualStrategyNumber)
+			continue;
+		if (boundary == GAPFILL_END && strategy != BTLessStrategyNumber &&
+			strategy != BTLessEqualStrategyNumber)
+			continue;
+
+		value = get_boundary_expr_value(state, boundary, expr);
+
+		/*
+		 * if the boundary expression operator does not match the operator
+		 * used by the gapfill node we adjust the value by 1 here
+		 *
+		 * the operators for the gapfill node are >= for start and < for end
+		 * column > value becomes start >= value + 1 column <= value becomes
+		 * end < value + 1
+		 */
+		if (strategy == BTGreaterStrategyNumber || strategy == BTLessEqualStrategyNumber)
+			value += 1;
+
+		if (!boundary_found)
+		{
+			boundary_found = true;
+			boundary_value = value;
+		}
+		else
+		{
+			if (boundary == GAPFILL_START)
+				boundary_value = Max(boundary_value, value);
+			else
+				boundary_value = Min(boundary_value, value);
+		}
+	}
+
+	if (boundary_found)
+		return boundary_value;
+
+	ereport(ERROR,
+			(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+			 errmsg("invalid time_bucket_gapfill argument: could not infer %s boundary from WHERE "
+					"clause",
+					boundary == GAPFILL_START ? "start" : "finish"),
+			 errhint("You can either pass start and finish as arguments or in the WHERE clause")));
+}
+
+static Const *
+make_const_value(Oid typid, int64 value)
```

Comment:
should be `make_const_value_for_gapfill_internal`

Reply:
changed